### PR TITLE
Add gated execution mode for Batch2

### PR DESCRIPTION
**Pull request by jamesagnew**

### CODE REVIEW

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/impl/JobCleanerServiceImpl.java [hunk 1]
```diff
@@ -255,16 +287,123 @@ private boolean updateInstanceStatus(JobInstance theInstance, StatusEnum newStat
...
+	 * are found, so that they can be reused for maintenance jobs without
+	 * needing to hit the database a second time.
+	 */
+	private static class JobChunkProgressAccumulator {
+
+		private final Set<String> myConsumedInstanceAndChunkIds = new HashSet<>();
+		private final Multimap<String, ChunkStatusCountKey> myInstanceIdToChunkStatuses = ArrayListMultimap.create();
+
+		public void addChunk(String theInstanceId, String theChunkId, String theStepId, StatusEnum theStatus) {
+			if (myConsumedInstanceAndChunkIds.add(theInstanceId + " " + theChunkId)) {
```
Comment (fil512) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/impl/JobCleanerServiceImpl.java:356]: question: why is this needed?

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/model/JobDefinition.java [hunk 2]
```diff
@@ -233,11 +254,42 @@ public <NPT extends IModelJson> Builder<NPT, NIT> setParametersType(@Nonnull Cla
...
+		 *    </li>
+		 * </ul>
+		 * </p>
+		 * <p>
+		 * Setting this mode means the job may take longer, since it will
+		 * rely on a polling mechanism to determine that one step is
+		 * complete before beginning any processing for the next step.
+		 * </p>
+		 */
+		public Builder<PT, NIT> gatedExecution() {
```
Comment (fil512) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/model/JobDefinition.java:290]: question: can this be an attribute of a step and not a job?  I could imagine a job with 7 steps and only the last one needs to be gated.

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/model/StatusEnum.java [hunk 3]
```diff
@@ -47,6 +49,17 @@ public enum StatusEnum {
...
 	 * result in a different outcome.
 	 */
-	FAILED
+	FAILED;
+
+	/**
+	 * Statuses that represent a job that has not yet completed. I.e.
+	 * all statuses except {@link #COMPLETED}
+	 */
+	public static final EnumSet<StatusEnum> INCOMPLETE_STATUSES = EnumSet.of(
```
Comment (fil512) [inline hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/model/StatusEnum.java:58]: suggestion: statically derive this from a property of StatusEnum rather than declaring it externally.  When declared externally like this, it runs the risk of future enums being added and accidentally omitted from this list.

File: hapi-fhir-storage-batch2/src/test/java/ca/uhn/fhir/batch2/impl/JobCleanerServiceImplTest.java [hunk 4]
```diff
@@ -47,26 +50,36 @@ public class JobCleanerServiceImplTest {
...
+	private IChannelProducer myWorkChannelProducer;
+	@Captor
+	private ArgumentCaptor<Message<JobWorkNotification>> myMessageCaptor;
 
 	@BeforeEach
 	public void beforeEach() {
-		mySvc = new JobCleanerServiceImpl(mySchedulerService, myJobPersistence);
+		myJobDefinitionRegistry = new JobDefinitionRegistry();
+		myBatchJobSender = new BatchJobSender(myWorkChannelProducer);
+		mySvc = new JobCleanerServiceImpl(mySchedulerService, myJobPersistence, myJobDefinitionRegistry, myBatchJobSender);
```
Comment (fil512) [inline hapi-fhir-storage-batch2/src/test/java/ca/uhn/fhir/batch2/impl/JobCleanerServiceImplTest.java:64]: suggestion: JobCleanerServiceImpl is probably not the right name any more...

**Review by fil512 — APPROVED** [review]

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/impl/JobCleanerServiceImpl.java [hunk 5]
```diff
@@ -255,16 +287,123 @@ private boolean updateInstanceStatus(JobInstance theInstance, StatusEnum newStat
 		return false;
 	}
 
+	private void triggerGatedExecutions(JobInstance theInstance, JobChunkProgressAccumulator theProgressAccumulator) {
+		if (theInstance.getStatus() != StatusEnum.IN_PROGRESS || theInstance.isCancelled()) {
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/impl/JobCleanerServiceImpl.java:291]: done

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/impl/JobCleanerServiceImpl.java [hunk 6]
```diff
@@ -255,16 +287,123 @@ private boolean updateInstanceStatus(JobInstance theInstance, StatusEnum newStat
...
+		@Nonnull
+		private Collection<ChunkStatusCountKey> getChunkStatuses(String theInstanceId) {
+			Collection<ChunkStatusCountKey> chunkStatuses = myInstanceIdToChunkStatuses.get(theInstanceId);
+			chunkStatuses = defaultIfNull(chunkStatuses, emptyList());
+			return chunkStatuses;
+		}
+
+
+		private static class ChunkStatusCountKey {
+			private final String myChunkId;
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/impl/JobCleanerServiceImpl.java:383]: Something about that pattern always looks so wrong to me... but i guess it's exactly what record types are, and we'll surely be using them when we get to JDK14 so might as well embrace it.

Done

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/impl/JobCleanerServiceImpl.java [hunk 7]
```diff
@@ -255,16 +287,123 @@ private boolean updateInstanceStatus(JobInstance theInstance, StatusEnum newStat
...
 	}
 
 
+	/**
+	 * While performing cleanup, the cleanup job loads all of the known
+	 * work chunks to examine their status. This bean collects the counts that
+	 * are found, so that they can be reused for maintenance jobs without
+	 * needing to hit the database a second time.
+	 */
+	private static class JobChunkProgressAccumulator {
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/impl/JobCleanerServiceImpl.java:350]: There are no tests for this class though, it is an internal implementation detail to the containing class. Disagree that it makes sense to increase visibility here.

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/impl/JobCleanerServiceImpl.java (hunk 1 above)
Comment (jamesagnew) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/impl/JobCleanerServiceImpl.java:356]: Adding javadocs:

```java
			// Note: If chunks are being written while we're executing, we may see the same chunk twice. This
			// check avoids adding it twice.
```

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/model/JobDefinition.java (hunk 2 above)
Comment (jamesagnew) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/model/JobDefinition.java:290]: Per discussion today in design meeting: This was deemed much harder for negligible benefit.

File: hapi-fhir-storage-batch2/src/test/java/ca/uhn/fhir/batch2/impl/JobCleanerServiceImplTest.java (hunk 4 above)
Comment (jamesagnew) [reply hapi-fhir-storage-batch2/src/test/java/ca/uhn/fhir/batch2/impl/JobCleanerServiceImplTest.java:64]: Good call, will rename

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/model/StatusEnum.java (hunk 3 above)
Comment (jamesagnew) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/model/StatusEnum.java:58]: done
